_posts/2025-09-05-anatomy-of-vllm.md

```diff
@@ -0,0 +1,992 @@
+* synchronous (all execution happens in a single blocking process)
+* single-GPU (no data/model/pipeline/expert parallelism; DP/TP/PP/EP = 1)
+* using standard transformer [[2]](#ref-2) (supporting hybrid models like Jamba requires a more complex hybrid KV-cache memory allocator)
+
+From here, we'll gradually build up to an online, async, multi-GPU, multi-node inference system - but still serving a standard transformer.
+
+In this example we do two things, we:
+
+1. Instantiate an engine
+3. Call <code>generate</code> on it to sample from the given prompts
+
+Let's start analyzing the constructor.
+
+### LLM Engine constructor
+The main components of the engine are:
+
+* vLLM config (contains all of the knobs for configuring model, cache, parallelism, etc.)
+* processor (turns raw inputs → <code>EngineCoreRequests</code> via validation, tokenization, and processing)
+* engine core client (in our running example we're using <code>InprocClient</code> which is basically == <code>EngineCore</code>; we'll gradually build up to <code>DPLBAsyncMPClient</code> which allows serving at scale)
+* output processor (converts raw <code>EngineCoreOutputs</code> → <code>RequestOutput</code> that the user sees)
+> [!NOTE]
+> With the V0 engine being deprecated, class names and details may shift. I'll emphasize the core ideas rather than exact signatures. I'll abstract away some but not all of those details.
+
+Engine core itself is made up of several sub components:
+
+* Model Executor (drives forward passes on the model, we're currently dealing with <code>UniProcExecutor</code> which has a single <code>Worker</code> process on a single GPU). We'll gradually build up to <code>MultiProcExecutor</code> which supports multiple GPUs
+* Structured Output Manager (used for guided decoding - we'll cover this later)
+* Scheduler (decides which requests go into the next engine step) - it further contains:
+  <ol type="a">
+  <li>policy setting - it can be either <b>FCFS</b> (first come first served) or <b>priority</b> (higher priority requests are served first)</li>
+  <li><code>waiting</code> and <code>running</code> queues</li>
+  <li>KV cache manager - the heart of paged attention [[3]](#ref-3)</li>
+
+The KV-cache manager maintains a <code>free_block_queue</code> - a pool of available KV-cache blocks (often on the order of hundreds of thousands, depending on VRAM size and block size). During paged attention, the blocks serve as the indexing structure that map tokens to their computed KV cache blocks.
+
+<p align="center">
+<picture>
+<img src="/assets/figures/2025-vllm-anatomy/engine_constructor.png" width="100%">
+</picture><br>
+<b>Figure 1</b>: Core components described in this section and their relationships
+</p>
+
+> [!NOTE]
+> Block size for a standard transformer layer (non-MLA [[4]](#ref-4)) is computed as follows:
+> 2 * <code>block_size</code> (default=16) * <code>num_kv_heads</code> * <code>head_size</code> * <code>dtype_num_bytes</code> (2 for bf16)
```

Comment:
the first `2` is for k and v right? `2 for bf16` is for `dtype_num_bytes`

Reply:
yep - clarified!